rust/src/lib/ifaces/ipvlan.rs

```diff
@@ -0,0 +1,114 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use serde::{Deserialize, Serialize};
+
+use crate::{BaseInterface, ErrorKind, InterfaceType, NmstateError};
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[non_exhaustive]
+/// Linux kernel IPVLAN interface. The example YAML output of
+/// [crate::NetworkState] with an IPVLAN interface would be:
+/// ```yaml
+/// ---
+/// interfaces:
+///   - name: ipvlan0
+///     type: ipvlan
+///     state: up
+///     ipvlan:
+///       base-iface: eth1
+///       mode: l3
+///       private: true
+/// ```
+pub struct IpVlanInterface {
+    #[serde(flatten)]
+    pub base: BaseInterface,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub ipvlan: Option<IpVlanConfig>,
+}
+
+impl Default for IpVlanInterface {
+    fn default() -> Self {
+        Self {
+            base: BaseInterface {
+                iface_type: InterfaceType::IpVlan,
+                ..Default::default()
+            },
+            ipvlan: None,
+        }
+    }
+}
+
+impl IpVlanInterface {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub(crate) fn sanitize(
+        &self,
+        is_desired: bool,
+    ) -> Result<(), NmstateError> {
+        if is_desired {
+            if let Some(conf) = &self.ipvlan {
+                if conf.private == Some(true) && conf.vepa == Some(true) {
+                    let e = NmstateError::new(
+                        ErrorKind::InvalidArgument,
+                        "Both private and VEPA flags cannot \
+                    be enabled at the same time"
+                            .to_string(),
+                    );
+                    log::error!("{}", e);
+                    return Err(e);
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Default)]
+#[serde(rename_all = "kebab-case", deny_unknown_fields)]
+#[non_exhaustive]
+pub struct IpVlanConfig {
+    pub base_iface: String,
+    pub mode: IpVlanMode,
```

Comment:
Both base-iface and mode can be optional when modifying existing interface. It also helps us on `gen_diff()`.

Reply:
Please also mark `mode` with `derive[serde(default)]`, user does not need to define it for default value.

Reply:
Sure! Let me adjust it, thanks!

Reply:
Done, thank you!